src/hooks/useNFTContract.js

```diff
@@ -0,0 +1,311 @@
+import {useEffect, useState, useContext} from 'react';
+import {ethers, BigNumber} from 'ethers';
+
+import {
+  CONTRACTS,
+} from './web3-constants.js';
+import {FTABI, NFTABI, WebaverseABI} from '../abis/contract.jsx';
+import {ChainContext} from './chainProvider.jsx';
+import {handleBlobUpload} from '../../util.js';
+import {registerLoad} from '../LoadingBox.jsx';
+
+const FILE_ADDRESS = 'https://ipfs.webaverse.com/ipfs/';
+
+// const CONTRACT_EVENTS = {
+//   MINT_COMPLETE: 'MintComplete',
+//   METADATA_SET: 'MetadataSet',
+//   SINGLE_METADATA_SET: 'SingleMetadataSet',
+//   HASH_UPDATE: 'HashUpdate',
+//   COLLABORATOR_ADDED: 'CollaboratorAdded',
+//   COLLABORATOR_REMOVED: 'CollaboratorRemoved',
+//   SINGLE_COLLABORATOR_ADDED: 'SingleCollaboratorAdded',
+//   SINGLE_COLLABORATOR_REMOVED: 'SingleCollaboratorRemoved',
+// };
+
+export default function useNFTContract(currentAccount) {
+  const {selectedChain} = useContext(ChainContext);
+  const [WebaversecontractAddress, setWebaversecontractAddress] = useState(null);
+  const [NFTcontractAddress, setNFTcontractAddress] = useState(null);
+  const [FTcontractAddress, setFTcontractAddress] = useState(null);
+
+  const [minting, setMinting] = useState(false);
+  const [showWallet, setShowWallet] = useState(false);
+  const [error, setError] = useState('');
+
+  useEffect(() => {
+    try {
+      if (CONTRACTS[selectedChain.contract_name]) {
+        const WebaversecontractAddress = CONTRACTS[selectedChain.contract_name].Webaverse;
+        const NFTcontractAddress = CONTRACTS[selectedChain.contract_name].NFT;
+        const FTcontractAddress = CONTRACTS[selectedChain.contract_name].FT;
+        setWebaversecontractAddress(WebaversecontractAddress);
+        setNFTcontractAddress(NFTcontractAddress);
+        setFTcontractAddress(FTcontractAddress);
+      } else {
+        throw new Error('unsupported chain');
+      }
+    } catch (error) {
+      console.warn('tried to use contract for unsupported chain');
+    }
+  }, [selectedChain]);
+
+  async function getSigner() {
+    var provider = new ethers.providers.Web3Provider(window.ethereum);
+    return provider.getSigner(currentAccount);
+  }
+
+  const getContract = async () => { // NFTcontract
+    const simpleRpcProvider = new ethers.providers.StaticJsonRpcProvider(selectedChain.rpcUrls[0]);
+    const contract = new ethers.Contract(NFTcontractAddress, NFTABI, simpleRpcProvider);
+    return contract;
+  };
+
+  async function mintNFT(currentApp, previewImage, callback = () => {}, afterminting = () => {}) {
```

Comment:
This entire code block is basically repeated in `https://github.com/webaverse/app/pull/3445/files#diff-4fe5770969890d344194764aa4b1778f5b6385e3aab7914b4ae69e9bac9a3bfeR166, so it should probably be better abstracted.

Reply:
mintNFT and mintfromVoucher are basically same.
But currently mintfromVoucher is for testing.
I will update this when serve drops feature and voucher server are completed.